pkg/virt-controller/watch/drain/disruptionbudget/disruptionbudget.go

```diff
@@ -314,44 +315,70 @@ func (c *DisruptionBudgetController) execute(key string) error {
 	}
 
 	// Only consider pdbs which belong to this vmi
-	pdb, err := c.pdbForVMI(namespace, name)
+	pdbs, err := c.pdbsForVMI(namespace, name)
 	if err != nil {
 		log.DefaultLogger().Reason(err).Error("Failed to fetch pod disruption budgets for namespace from cache.")
 		// If the situation does not change there is no benefit in retrying
 		return nil
 	}
 
-	return c.sync(key, vmi, pdb)
+	if len(pdbs) == 0 {
+		return c.sync(key, vmi, nil)
+	}
 
+	for i := range pdbs {
+		if syncErr := c.sync(key, vmi, pdbs[i]); syncErr != nil {
+			err = syncErr
+		}
+	}
+	return err
 }
 
-func (c *DisruptionBudgetController) sync(key string, vmi *virtv1.VirtualMachineInstance, pdb *v1beta1.PodDisruptionBudget) error {
-	delete := false
-	create := false
-	// delete if there is no VMI
-	if vmi == nil && pdb != nil {
-		delete = true
-	} else if vmi != nil {
-		wantsToMigrate := wantsToMigrateOnDrain(vmi)
-		if vmi.DeletionTimestamp != nil && pdb != nil {
-			// pdb can already be deleted, shutdown already in process
-			delete = true
-		} else if !wantsToMigrate && pdb != nil {
-			// We don't want migrations on evictions, if there is a pdb, remove it
-			delete = true
-		} else if wantsToMigrate && vmi.DeletionTimestamp == nil && pdb == nil {
-			// No pdb and we want migrations on evictions
-			create = true
-		} else if wantsToMigrate && pdb != nil {
-			if ownerRef := v1.GetControllerOf(pdb); ownerRef != nil && ownerRef.UID != vmi.UID {
-				// The pdb is from an old vmi with a different uid, delete and later create the correct one
-				// The VMI always has a minimum grace period, so normally this should not happen, therefore no optimizations
-				delete = true
-			}
+func (c *DisruptionBudgetController) isMigrationComplete(vmi *virtv1.VirtualMachineInstance, migrationName string) (bool, error) {
+	notFoundErrorCheck := func(err error) (bool, error) {
+		if errors.IsNotFound(err) {
+			return true, nil
 		}
+		return false, err
+	}
+
+	vmim, err := c.clientset.VirtualMachineInstanceMigration(vmi.Namespace).Get(migrationName, &v1.GetOptions{})
+	if err != nil {
+		return notFoundErrorCheck(err)
 	}
 
-	if delete && pdb != nil && pdb.DeletionTimestamp == nil {
+	if vmim.IsFinal() {
```

Comment:
check the VMI state too. 

```
if vmim.IsFinal() ||
   (vmi.Status.MigrationState.MigrationUID == string(vmim.UID) && vmi.Status.MigrationState.Completed) {

    return true
}
```

The only reason to check the VMI is that there' s a lag between when the vmim reports success and when the vmi first confirms the migration has completed. Since we're not watching for VMIM objects in thsi controller, we'd need to check both to ensure the gap closes.

Reply:
Done!